### PR TITLE
Taiko SR/pp system rework

### PR DESCRIPTION
**Pull request by chayleaf**
Removed rhythm bonus, improved pattern complexity system, changed speed bonus, rebalanced values to be more similar to std's pp system. See code for details?

### CODE REVIEW

**Comment by pavlukivan** [comment]
An issue I found is Towa yori Towa ni being almost 6*, I'll try looking into it.

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs [hunk 1]
```diff
@@ -11,85 +13,101 @@ namespace osu.Game.Rulesets.Taiko.Difficulty.Skills
...
-
-            lastColourSwitch = newColourSwitch;
-            sameColourCount = 1;
-
-            // We only want a bonus if the parity of the color switch changes
-            return oldColourSwitch != ColourSwitch.None && oldColourSwitch != newColourSwitch;
+            // Sliders and spinners are optional to hit and thus are ignored
+            if (!(current.LastObject is Hit))
+				return 0.0;
+			// Decay known patterns
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs]: Missing new line after `return`

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs [hunk 2]
```diff
@@ -11,85 +13,101 @@ namespace osu.Game.Rulesets.Taiko.Difficulty.Skills
...
+                return 0.0;
+            // Decay known patterns
+            noteNum++;
+
+            // Arbitrary values found by analyzing top plays of different people
+            double addition = 1.9 + 2.5 * colourChangeDifficulty(current);
+
+            double deltaSum = current.DeltaTime;
+            int deltaCount = 1;
+            for(var i = 1; i < maxPatternLength; i++)
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs]: Missing space after statements

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs [hunk 3]
```diff
@@ -11,85 +13,101 @@ namespace osu.Game.Rulesets.Taiko.Difficulty.Skills
...
-                if (hasRhythmChange(current))
-                    addition += 1;
-            }
-            else
+        protected override double StrainValueOf(DifficultyHitObject current)
+        {
+            // Sliders and spinners are optional to hit and thus are ignored
+            if (!(current.LastObject is Hit))
+                return 0.0;
+            // Decay known patterns
```
Comment (iiSaLMaN) [inline osu.Game.Rulesets.Taiko/Difficulty/Skills/Strain.cs]: Missing newline after return

**Review by frenzibyte — CHANGES_REQUESTED** [review]

**Comment by pavlukivan** [comment]
idk what to do about the other 2 warnings, i can't remove the initializer because otherwise the elements will be null

**Comment by peppy** [comment]
Thanks for the contribution. Do you have any external reading or statistical analysis to go with these changes (on forum or elsewhere)?

**Comment by pavlukivan** [comment]
I will do some analysis soon, and the changes themselves are better described with code, the only base for numbers is manual play analyzing so far, now I'm trying to split speed and reading into different skills

**Comment by pavlukivan** [comment]
SR difference table: https://docs.google.com/spreadsheets/d/1gn-FcSf2Vm0jaVMPFPYz2CYlz-__zI1JNVfNQr0DmkM/edit?usp=sharing
Note that this is slightly outdated (I fixed some maps being NaN/Inf, now 0 spacing is weighted normally and negative spacing gives like 100-200 stars).

**Comment by pavlukivan** [comment]
what is that appveyor warning even?

**Comment by iiSaLMaN** [comment]
> what is that appveyor warning even?

That's not a warning (error you meant), it's a test that failed due to your changes for the taiko difficulty calculator: go to the build, check the `Tests` tab

**Comment by pavlukivan** [comment]
[Updated SR table](https://docs.google.com/spreadsheets/d/1wgQl2E7QSbddnllpK4uQUkBXxfy5N5r34PlryEBwx04/edit?usp=sharing)

### Issues

Maps with negative delta time are broken, and i really don't want to fix it because the fact negative delta times even exist is a bug. Another issue is all ranked 0\* maps being buffed to 1\*, because of having 1\* reading and about 0.05\* speed, but I'm not sure I really want to fix that.

Ney has converts rated highly in his top plays, I'm not sure if that's something that's supposed to happen because I didn't check the maps, it needs to be looked into, in the worst case scenario converts need to be unranked.

Waiting for a test server to go up to see if anything is heavily broken, overall I'm satisfied with the new ratings, even though some maps end up being a bit over/under ranked.

### Notes for the future

Tech maps are now heavily underrated because the reason they had high SR before is rhythm changes being overrated, which used to offset the low rating otherwise. That's something go be considered in future iterations, SV needs to be accounted for, because both low SV and high SV are hard, relatively to note density. In std AR doesn't affect SR, so we probably need to do the same, and it needs to be a part of both strain pp and acc pp (low sv is harder to acc because you need to rely on song more).

Rhythm changes still need to be looked into, but as a part of accuracy pp, not star rating.

**Comment by pavlukivan** [comment]
I did nerf farmy converts, however it comes at a cost - some ranked maps lose too much value, especially Thank You Friends (because the hard repeats 3 times without changes). Converts are like a completely different playstyle, because TL beating allows you to play streams that are 2x as fast, or even more. I think accounting for them hurts the pp system, so they should be unranked.

**Review by cory2067 — COMMENTED** [review]
the Reading skill is hard to understand -- but im pretty on board with how the results look
having a speed skill is cool, but idk if i agree with completely doing away with the rhythm bonus

for example here are some results, comparing new and old SR on various patterns (185bpm for 256/1)
New: https://paste.ee/p/O1PMJ
Old: https://paste.ee/p/euNN8

from the above data, kkdd 1/4 stream is 3.9 stars, whereas kk dd 1/8 doubles are 3.14
i'd figure that 1/8 doubles are at least as hard as the equivalent 1/4 pattern, not easier
it's the same notes/second, but it's harder to execute because of the non-even rhythm (i would argue)

i'd be curious to see what other people think about that

**Comment by cory2067** [comment]
> I did nerf farmy converts, however it comes at a cost - some ranked maps lose too much value, especially Thank You Friends (because the hard part repeats 3 times without changes).

wait, two of the diff spikes in thank you friends are separated by like 2000 combo
does the algorithm really remember (and nerf) patterns from that long ago?

**Comment by pavlukivan** [comment]
Repeating this about the rhythm bonus - 1/8 patterns in streams are harder than 1/6, yet they have no rhythm complexity. And in Thank You Friends the problem is the stream is the only part with hard patterns, so whenever it sees that chunk it knows what's coming next.

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+                previousDeltas[i] = previousDeltas[i - 1];
+            }
+
+            previousDeltas[0] = current.DeltaTime;
+
+            // Use last N notes instead of last 1 note for determining pattern speed. Especially affects 1/8 doubles.
+            var normalizedDelta = deltaSum / deltaCount;
+
+            // Overwrite current.DeltaTime with normalizedDelta in Skill's strainDecay function
```
[the remark this answers is not in the document]
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs]: Look at `osu.Game.Rulesets.Difficulty.Skills.Skill.strainDecay`. I inverse that function to replace the function output. Kinda hacky, but I think it's better than writing a completely different skill class.

[the remark this answers is not in the document]
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs]: Look at `osu.Game.Rulesets.Difficulty.Skills.Skill.strainDecay`. I inverse that function to replace the function output. Kinda hacky, but I think it's better than writing a completely different skill class.

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+                // How well is the pattern remembered
+                double[] memory = new double[2];
+                int[] count = new int[2];
+
+                for (var j = 0; j < 2; j++)
+                {
+                    int n;
+                    patternCount[j].TryGetValue(pattern, out count[j]);
+                    patternOccur[j].TryGetValue(pattern, out n);
```
[the remark this answers is not in the document]
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs]: Patterns are sorted to 2 dictionaries by last note, it makes some stuff faster I guess? If it doesn't, then it's pretty useless and can be replaced, but in that case you would need to manually add first `d`, then `k` to the patterns to check their rating.

File: osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+        // Pattern occurences. An optimization is possible using bitarrays instead of strings.
+        private readonly Dictionary<string, int>[] patternOccur = { new Dictionary<string, int>(), new Dictionary<string, int>() };
+        private readonly Dictionary<string, int>[] patternCount = { new Dictionary<string, int>(), new Dictionary<string, int>() };
+
+        private readonly double[] previousDeltas = new double[max_pattern_length];
+        private int noteNum;
+
+        private const double rhythm_change_base_threshold = 0.2;
+        private const double rhythm_change_base = 2.0;
```
[the remark this answers is not in the document]
Comment (pavlukivan) [reply osu.Game.Rulesets.Taiko/Difficulty/Skills/Reading.cs]: Will do

**Comment by pavlukivan** [comment]
Really proud of this one! Diff spikes nerfed successfully.

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs [hunk 7]
```diff
@@ -68,28 +68,39 @@ public override double Calculate(Dictionary<string, double> categoryDifficulty =
 
         private double computeStrainValue()
         {
-            double strainValue = Math.Pow(5.0 * Math.Max(1.0, Attributes.StarRating / 0.0075) - 4.0, 2.0) / 100000.0;
+            double strainValue = Math.Pow(Math.Max(1.0, Attributes.StarRating / 0.0075) - 4.0, 2.53) / 100000.0;
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs]: The `pow(x, 2.53)` results in an imaginary number if x < 0.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
Going to close this for now as it is quite old and there has been further work on SR since (https://github.com/ppy/osu/pull/9294). Thanks!

**Pull request closed without merging**